timelock-corruption-detection/src/main/java/com/palantir/timelock/history/PaxosLogHistoryProgressTracker.java

```diff
@@ -0,0 +1,112 @@
+        this.sqlitePaxosStateLogHistory = sqlitePaxosStateLogHistory;
+    }
+
+    public SequenceBounds getPaxosLogSequenceBounds(NamespaceAndUseCase namespaceAndUseCase) {
+        ProgressComponents progress = getOrPopulateProgressComponents(namespaceAndUseCase);
+        return SequenceBounds.builder()
+                .lower(progress.seq())
+                .upper(progress.seq() + MAX_ROWS_ALLOWED)
+                .build();
+    }
+
+    public void updateProgressState(Map<NamespaceAndUseCase, SequenceBounds> namespaceAndUseCaseSequenceBoundsMap) {
+        namespaceAndUseCaseSequenceBoundsMap.forEach((namespaceAndUseCase, bounds) ->
+                updateProgressStateForNamespaceAndUseCase(namespaceAndUseCase, bounds));
+    }
+
+    private ProgressComponents getOrPopulateProgressComponents(NamespaceAndUseCase namespaceAndUseCase) {
+        return verificationProgressStateCache.computeIfAbsent(namespaceAndUseCase, this::getLastVerifiedSeqFromLogs);
+    }
+
+    private ProgressComponents getLastVerifiedSeqFromLogs(NamespaceAndUseCase namespaceAndUseCase) {
+        Client client = namespaceAndUseCase.namespace();
+        String useCase = namespaceAndUseCase.useCase();
+
+        return logVerificationProgressState
+                .getProgressComponents(client, useCase)
+                .orElseGet(() -> logVerificationProgressState.resetProgressState(
+                        client, useCase, getLatestSequenceForNamespaceAndUseCase(namespaceAndUseCase)));
+    }
+
+    @VisibleForTesting
+    void updateProgressStateForNamespaceAndUseCase(NamespaceAndUseCase key, SequenceBounds value) {
+        long lastVerifiedSequence = value.upper();
+
+        ProgressComponents currentProgressState = getOrPopulateProgressComponents(key);
+        resetIfRequired(key, value, currentProgressState)
+                .orElseGet(
+                        () -> updateProgressInDbThroughCache(key, value, lastVerifiedSequence, currentProgressState));
+    }
+
+    private ProgressComponents updateProgressInDbThroughCache(
+            NamespaceAndUseCase key,
+            SequenceBounds value,
+            long lastVerifiedSequence,
+            ProgressComponents progressComponents) {
+        ProgressComponents progressState = ProgressComponents.builder()
+                .progressLimit(progressComponents.progressLimit())
+                .seq(value.upper())
+                .build();
+        verificationProgressStateCache.put(key, progressState);
+        logVerificationProgressState.updateProgress(key.namespace(), key.useCase(), lastVerifiedSequence);
+        return progressState;
+    }
+
+    private Optional<ProgressComponents> resetIfRequired(
+            NamespaceAndUseCase key, SequenceBounds value, ProgressComponents currentProgressState) {
+        if (value.upper() <= currentProgressState.progressLimit()) {
+            return Optional.empty();
+        }
```

Comment:
This doesn't behave the way I think is intended: your seeding of the bound in `getLatestSequence` is inclusive, however your use of `value` here is exclusive. This means that even if there are no new entries, the validation task will repeatedly cycle over the entries (e.g. of an inactive client), which I doubt is intended.